### PR TITLE
neovim: allow extra Lua packages

### PR DESCRIPTION
**Pull request by gpanders**
###### Motivation for this change

Users can already specify additional Python, Ruby, and Node. This change allows
users to specify Lua packages as well.

I am marking this as a draft because I would love some input from other maintainers. I am not sure this is the best way to implement this, though I did the best I could with the Nix knowledge I have. In particular, I don't like that the `lua` parameter now has to be specified for both `neovim-unwrapped` as well as `neovimUtils`, but I'm not sure how to get around that.

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [ ] aarch64-darwin
- [x] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/editors/neovim/utils.nix [hunk 1]
```diff
@@ -110,6 +114,10 @@ let
           "--set" "GEM_HOME" "${rubyEnv}/${rubyEnv.ruby.gemPath}"
         ] ++ lib.optionals (binPath != "") [
           "--suffix" "PATH" ":" binPath
+        ] ++ lib.optionals (luaEnv != null) [
+          "--prefix" "LUA_PATH" ";" "${luaEnv}/share/lua/${lua.luaversion}/?.lua"
+        ] ++ lib.optionals (luaEnv != null) [
```
Comment (teto) [inline pkgs/applications/editors/neovim/utils.nix:119]: you can merge these two

**Comment by teto** [comment]
Looks useful, could you add a test in pkgs/applications/editors/neovim/tests.nix ? (check that a specified lua package was loaded for instance)

**Comment by gpanders** [comment]
Does the test look okay?

File: pkgs/applications/editors/neovim/tests.nix [hunk 2]
```diff
@@ -132,4 +139,6 @@ rec {
     extraName = "-pathogen";
     configure.pathogen.pluginNames = [ "vim-nix" ];
   };
+
+  test_nvim_with_lua_packages = wrapNeovim2 "-with-lua-packages" nvimWithLuaPackages;
```
Comment (teto) [inline pkgs/applications/editors/neovim/tests.nix:143]: the test should start neovim to check if the require works.

Comment (gpanders) [reply pkgs/applications/editors/neovim/tests.nix:143]: Done

**Comment by gpanders** [comment]
I also added a `lua` attribute to `neovim-unwrapped` so that the wrapper can just use that value. This removes the need to duplicate the `lua` parameter in `all-packages.nix`.

**Comment by teto** [comment]
@GrahamcOfBorg test neovim

**Comment by teto** [comment]
@GrahamcOfBorg test neovim-unwrapped

File: pkgs/applications/editors/neovim/default.nix [hunk 3]
```diff
@@ -50,6 +50,8 @@ in
 
     dontFixCmake = true;
 
+    inherit lua;
```
Comment (teto) [inline pkgs/applications/editors/neovim/default.nix:53]: This type of thing usually go in "passthru" but looks like it works.

**Comment by teto** [comment]
thanks for contributing to upstream neovim and now nixpkgs. Much appreciated.

**Pull request merged**